src/components/geolocate-control.js

```diff
@@ -0,0 +1,224 @@
+      this._mapboxGeolocateControl._geolocationWatchID = undefined;
+    }
+  }
+
+  _setupMapboxGeolocateControl = (supportsGeolocation) => {
+    if (!supportsGeolocation) {
+      /* eslint-disable no-console, no-undef */
+      console.warn(
+        'Geolocation support is not available, the GeolocateControl will not be visible.'
+      );
+      /* eslint-enable no-console, no-undef */
+      return;
+    }
+
+    const controlOptions = {};
+    ['positionOptions', 'fitBoundsOptions', 'trackUserLocation', 'showUserLocation'].forEach(
+      prop => {
+        // For null option, use Mapbox default value
+        if (prop in this.props && this.props[prop] !== null) {
+          controlOptions[prop] = this.props[prop];
+        }
+      }
+    );
+
+    this._mapboxGeolocateControl = new MapboxGeolocateControl(controlOptions);
+
+    // the following re-implement MapboxGeolocateControl's _setupUI
+    // replace mapbox internal prop
+    this._mapboxGeolocateControl._watchState = 'OFF';
+
+    // replace mapbox internal UI elements
+    this._mapboxGeolocateControl._geolocateButton = this._geolocateButtonRef.current;
+
+    // replace mapbox internal methods
+    this._mapboxGeolocateControl._updateMarker = this._updateMarker;
+    this._mapboxGeolocateControl._updateCamera = this._updateCamera;
+
+    this._mapboxGeolocateControl._setup = true;
+  };
+
+  _onClickGeolocate = () => {
+    return this._mapboxGeolocateControl.trigger();
+  };
+
+  _updateMarker = position => {
+    if (position) {
+      this.setState({markerPosition: position.coords});
+    }
+  };
+
+  _getBounds = position => {
+    const center = new LngLat(
+      position.coords.longitude,
+      position.coords.latitude
+    );
+    const radius = position.coords.accuracy;
+    const bounds = center.toBounds(radius);
+
+    return [
+      [bounds._ne.lng, bounds._ne.lat],
+      [bounds._sw.lng, bounds._sw.lat]
+    ];
+  };
+
+  _updateCamera = position => {
+    const {viewport} = this._context;
+
+    const bounds = this._getBounds(position);
+    const {longitude, latitude, zoom} = new WebMercatorViewport(viewport).fitBounds(bounds);
+
+    const newViewState = Object.assign({}, viewport, {longitude, latitude, zoom});
+    const mapState = new MapState(newViewState);
+    const viewState = Object.assign(
+      {},
+      mapState.getViewportProps(),
+      LINEAR_TRANSITION_PROPS
+    );
+
+    // Call new style callback
+    this.props.onViewStateChange({viewState});
+
+    // Call old style callback
+    this.props.onViewportChange(viewState);
+  };
+
+  _renderButton = (type, label, callback, children) => {
+    return createElement('button', {
+      key: type,
+      className: `mapboxgl-ctrl-icon mapboxgl-ctrl-${type}`,
+      ref: this._geolocateButtonRef,
+      type: 'button',
+      title: label,
+      onClick: callback,
+      children
+    });
+  };
+
+  _renderMarker = () => {
+    const {markerPosition} = this.state;
+    if (!markerPosition) {
```

Comment:
Need to check `showUserLocation` too?

Reply:
checked in the `render()`.

Maybe I could move that logic to `_renderMarker`.